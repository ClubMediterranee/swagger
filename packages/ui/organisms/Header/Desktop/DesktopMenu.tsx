"use client";

import { ButtonContent } from "@clubmed/trident-ui/molecules/Buttons/ButtonContent";
import { useButton } from "@clubmed/trident-ui/molecules/Buttons/v2/Button";
import { getButtonClasses } from "@clubmed/trident-ui/molecules/Buttons/v2/Button.type";
import classnames from "classnames";
import {
  Children,
  type ComponentPropsWithoutRef,
  type FunctionComponent,
  type PropsWithChildren,
  type ReactNode,
  useEffect,
  useId,
  useRef,
  useState
} from "react";

import type { NavItem } from "../types/NavItem";

export const HOVER_ENTRY_DURATION = 300;
export const HOVER_EXIT_DURATION = 500;

type SetMenu = (index: number) => void;
type ResetMenu = () => void;

interface Props extends Omit<ComponentPropsWithoutRef<"nav">, "children"> {
  children: (props: { activeItem: number; resetMenu: ResetMenu; setMenu: SetMenu }) => ReactNode;
}

export const DesktopMenu: FunctionComponent<Props> = ({ children, className, ...attrs }) => {
  const hoverInTimeout = useRef<ReturnType<typeof setTimeout>>(0 as any);
  const hoverOutTimeout = useRef<ReturnType<typeof setTimeout>>(0 as any);

  const [activeItem, setActiveItem] = useState(-1);

  useEffect(() => {
    return () => {
      clearTimeout(hoverOutTimeout.current);
      clearTimeout(hoverInTimeout.current);
    };
  }, []);

  const setMenu: SetMenu = (index) => {
    clearTimeout(hoverOutTimeout.current);
    hoverInTimeout.current = setTimeout(() => {
      setActiveItem(index);
    }, HOVER_ENTRY_DURATION);
  };

  const resetMenu: ResetMenu = () => {
    clearTimeout(hoverInTimeout.current);
    hoverOutTimeout.current = setTimeout(() => {
      setActiveItem(-1);
    }, HOVER_EXIT_DURATION);
  };

  return (
    <nav {...attrs} className={classnames("hidden items-center gap-x-12 px-8 lg:flex", className)} role="navigation">
      {children({ activeItem, setMenu, resetMenu })}
    </nav>
  );
};

const ButtonAnchor: FunctionComponent<any> = ({
  className,
  children,
  color = "saffron",
  icon,
  iconWidth,
  size = "medium",
  theme = "solid",
  variant = "pill",
  component: Link,
  ...props
}) => {
  const { attrs } = useButton(props);

  return (
    <Link {...attrs} className={classnames(getButtonClasses({ color, size, theme, variant }), className)}>
      <ButtonContent icon={icon} iconWidth={iconWidth}>
        {children}
      </ButtonContent>
    </Link>
  );
};

interface DesktopMenuItemProps {
  activeItem: number;
  index: number;
  Link?: any;
  item: NavItem;
  onClick?: () => void;
  resetMenu: ResetMenu;
  setMenu: SetMenu;
}

export const DesktopMenuItem: FunctionComponent<PropsWithChildren<DesktopMenuItemProps>> = ({
  activeItem,
  children,
  index,
  item,
  onClick,
  resetMenu,
  Link = "a",
  setMenu
}) => {
  const btnId = useId();
  const menuId = useId();
  const hasChildren = Children.count(children) > 0;

  return (
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <div
      onMouseEnter={() => {
        setMenu(index);
      }}
      onMouseLeave={() => {
        resetMenu();
      }}
    >
      <ButtonAnchor
        component={Link}
        aria-controls={hasChildren ? menuId : undefined}
        aria-haspopup={hasChildren}
        className={classnames("!border-0", {
          "cursor-pointer": Boolean(item.url),
          "cursor-default": !item.url
        })}
        color="black"
        href={item.url === "" ? "#" : item.url || "#"}
        rel={item.external ? "external" : undefined}
        target={item.external ? "_blank" : undefined}
        id={btnId}
        onClick={onClick}
        theme="outline"
      >
        {item.label}
      </ButtonAnchor>
      {hasChildren && (
        <div
          className={classnames("absolute inset-x-0 top-full bg-white px-20 py-40", {
            hidden: index !== activeItem
          })}
        >
          <div aria-labelledby={btnId} className="mb-24 flex columns-5 justify-center gap-x-40" id={menuId} role="menu">
            {children}
          </div>
        </div>
      )}
    </div>
  );
};

export const DesktopMenuItemColumn: FunctionComponent<PropsWithChildren> = ({ children }) => {
  return <div className="break-inside-avoid-column space-y-32 text-b3">{children}</div>;
};

type DesktopMenuSectionProps = NavItem["columns"][number]["sections"][number] & {
  Link?: any;
  onClick?: () => void;
};

export const DesktopMenuSection: FunctionComponent<PropsWithChildren<DesktopMenuSectionProps>> = ({
  children,
  onClick,
  label,
  url,
  external,
  Link = "a"
}) => {
  const hasLinks = Children.count(children) > 0;

  return (
    <div role="group">
      {label && (
        <Link
          className="block font-bold"
          href={url || undefined}
          onClick={onClick}
          role="menuitem"
          rel={external ? "external" : undefined}
          target={external ? "_blank" : undefined}
        >
          {label}
        </Link>
      )}
      {hasLinks && (
        <ul className="mt-20 space-y-8" role="group">
          {children}
        </ul>
      )}
    </div>
  );
};

interface DesktopMenuLinkProps {
  link: DesktopMenuSectionProps["links"][number];
  onClick?: () => void;
  Link?: any;
}

export const DesktopMenuLink: FunctionComponent<DesktopMenuLinkProps> = ({ link, onClick, Link = "a" }) => {
  return (
    <li role="presentation">
      <Link
        href={link.url || undefined}
        onClick={onClick}
        rel={link.external ? "external" : undefined}
        target={link.external ? "_blank" : undefined}
        role="menuitem"
      >
        {link.label}
      </Link>
    </li>
  );
};
