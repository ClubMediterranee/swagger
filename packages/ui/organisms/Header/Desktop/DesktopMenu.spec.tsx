import { fireEvent, render, screen, waitFor } from "@testing-library/react";

import type { NavItem } from "../types/NavItem";
import { DesktopMenu, DesktopMenuItem, DesktopMenuItemColumn, DesktopMenuLink, DesktopMenuSection } from "./DesktopMenu.js";

const links = [
  {
    label: "All inclusive sun holidays",
    url: "/o/all-inclusive-sun-holidays"
  },
  {
    label: "All Inclusive ski holidays",
    url: "/o/all-inclusive-ski-holidays"
  },
  {
    label: "Ski comparator",
    url: "/l/ski-comparator"
  },
  {
    label: "Organized travel tours",
    url: "/l/organized-travel-tours"
  }
];

const sections = [
  {
    label: "Our products",
    url: "/s",
    links
  }
];

const items = [
  {
    label: "Discover",
    url: "/l/discover-club-med",
    columns: [
      {
        sections
      }
    ]
  }
] satisfies NavItem[];

describe("<DesktopMenu />", () => {
  it("renders DesktopMenu", () => {
    const { container } = render(
      <DesktopMenu>
        {(props) => (
          <DesktopMenuItem item={items[0]} index={0} {...props}>
            item
          </DesktopMenuItem>
        )}
      </DesktopMenu>
    );
    expect(container).toMatchSnapshot("DesktopMenu");
  });

  describe("renders DesktopMenuItem", () => {
    it("opens and closes the menu", async () => {
      render(
        <DesktopMenu>
          {(props) => (
            <DesktopMenuItem item={items[0]} index={0} {...props}>
              item
            </DesktopMenuItem>
          )}
        </DesktopMenu>
      );

      expect(screen.getByRole("menu").parentElement?.getAttribute("class")).toContain("hidden");

      const trigger = screen.getByRole("link", { name: "Discover" });

      fireEvent.mouseEnter(trigger);

      await waitFor(() => {
        expect(screen.getByRole("menu").parentElement?.getAttribute("class")).not.toContain("hidden");
      });

      fireEvent.mouseLeave(trigger);

      await waitFor(() => {
        expect(screen.getByRole("menu").parentElement?.getAttribute("class")).toContain("hidden");
      });
    });

    it("calls onClick callback on opening", () => {
      const onClick = vi.fn();

      render(
        <DesktopMenu>
          {(props) => (
            <DesktopMenuItem item={items[0]} index={0} {...props} onClick={onClick}>
              item
            </DesktopMenuItem>
          )}
        </DesktopMenu>
      );

      const trigger = screen.getByRole("link", { name: "Discover" });

      fireEvent.click(trigger);

      expect(onClick).toHaveBeenCalledTimes(1);
    });

    it("renders a link without href when url is empty string", () => {
      render(
        <DesktopMenu>
          {(props) => (
            <DesktopMenuItem item={{ ...items[0], url: "" }} index={0} {...props}>
              item
            </DesktopMenuItem>
          )}
        </DesktopMenu>
      );

      const MenuItem = screen.getByRole("link", { name: "Discover" });

      expect(MenuItem.getAttribute("href")).toBe("#");
    });
  });

  it("renders DesktopMenuItemColumn", () => {
    const { container } = render(
      <DesktopMenu>
        {(props) => (
          <DesktopMenuItem item={items[0]} index={0} {...props}>
            <DesktopMenuItemColumn>Column</DesktopMenuItemColumn>
          </DesktopMenuItem>
        )}
      </DesktopMenu>
    );
    expect(container).toMatchSnapshot("DesktopMenuItemColumn");
  });

  describe("<DesktopMenuItemSection />", () => {
    it("renders DesktopMenuItemSection & DesktopMenuItemLink", () => {
      const { container } = render(
        <DesktopMenu>
          {(props) => (
            <DesktopMenuItem item={items[0]} index={0} {...props}>
              <DesktopMenuItemColumn>
                <DesktopMenuSection {...items[0].columns[0].sections[0]}>
                  <DesktopMenuLink link={links[0]} />
                </DesktopMenuSection>
              </DesktopMenuItemColumn>
            </DesktopMenuItem>
          )}
        </DesktopMenu>
      );
      expect(container).toMatchSnapshot("DesktopMenuItemColumn");
    });

    it("renders DesktopMenuItemSection without href when url is empty string", () => {
      render(
        <DesktopMenu>
          {(props) => (
            <DesktopMenuItem item={items[0]} index={0} {...props}>
              <DesktopMenuItemColumn>
                <DesktopMenuSection {...{ ...items[0].columns[0].sections[0], url: "" }}>
                  <DesktopMenuLink link={links[0]} />
                </DesktopMenuSection>
              </DesktopMenuItemColumn>
            </DesktopMenuItem>
          )}
        </DesktopMenu>
      );

      const trigger = screen.getByText("Our products");

      expect(trigger).not.toHaveAttribute("href");
    });

    it("calls onClick callback when clicked", () => {
      const onClick = vi.fn();

      render(
        <DesktopMenu>
          {(props) => (
            <DesktopMenuItem item={items[0]} index={0} {...props}>
              <DesktopMenuItemColumn>
                <DesktopMenuSection onClick={onClick} {...items[0].columns[0].sections[0]}>
                  <DesktopMenuLink link={links[0]} />
                </DesktopMenuSection>
              </DesktopMenuItemColumn>
            </DesktopMenuItem>
          )}
        </DesktopMenu>
      );

      const trigger = screen.getByRole("menuitem", { name: "Our products" });

      fireEvent.click(trigger);

      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });

  describe("<DesktopMenuItemLink />", () => {
    it("renders without href when url is empty string", () => {
      const { container } = render(
        <DesktopMenu>
          {(props) => (
            <DesktopMenuItem item={items[0]} index={0} {...props}>
              <DesktopMenuItemColumn>
                <DesktopMenuSection {...items[0].columns[0].sections[0]}>
                  <DesktopMenuLink link={{ ...links[0], url: "" }} />
                </DesktopMenuSection>
              </DesktopMenuItemColumn>
            </DesktopMenuItem>
          )}
        </DesktopMenu>
      );

      expect(container).toMatchSnapshot("DesktopMenuItemLink with empty url");
    });

    it("calls onClick callback when clicked", () => {
      const onClick = vi.fn();

      render(
        <DesktopMenu>
          {(props) => (
            <DesktopMenuItem item={items[0]} index={0} {...props}>
              <DesktopMenuItemColumn>
                <DesktopMenuSection {...items[0].columns[0].sections[0]}>
                  <DesktopMenuLink link={links[0]} onClick={onClick} />
                </DesktopMenuSection>
              </DesktopMenuItemColumn>
            </DesktopMenuItem>
          )}
        </DesktopMenu>
      );

      const trigger = screen.getByRole("menuitem", { name: "All inclusive sun holidays" });

      fireEvent.click(trigger);

      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });
});
