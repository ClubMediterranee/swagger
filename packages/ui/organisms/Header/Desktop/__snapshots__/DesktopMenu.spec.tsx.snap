// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<DesktopMenu /> > <DesktopMenuItemLink /> > renders without href when url is empty string > DesktopMenuItemLink with empty url 1`] = `
<div>
  <nav
    class="hidden items-center gap-x-12 px-8 lg:flex"
    role="navigation"
  >
    <div>
      <a
        aria-controls=":rh:"
        aria-haspopup="true"
        class="flex button button-outline button-black button-pill button-medium !border-0 cursor-pointer"
        data-name="ButtonAnchor"
        href="/l/discover-club-med"
        id=":rg:"
      >
        <span
          class="flex items-center justify-center gap-x-8 w-full"
        >
          Discover
        </span>
      </a>
      <div
        class="absolute inset-x-0 top-full bg-white px-20 py-40 hidden"
      >
        <div
          aria-labelledby=":rg:"
          class="mb-24 flex columns-5 justify-center gap-x-40"
          id=":rh:"
          role="menu"
        >
          <div
            class="break-inside-avoid-column space-y-32 text-b3"
          >
            <div
              role="group"
            >
              <a
                class="block font-bold"
                href="/s"
                role="menuitem"
              >
                Our products
              </a>
              <ul
                class="mt-20 space-y-8"
                role="group"
              >
                <li
                  role="presentation"
                >
                  <a
                    role="menuitem"
                  >
                    All inclusive sun holidays
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;

exports[`<DesktopMenu /> > <DesktopMenuItemSection /> > renders DesktopMenuItemSection & DesktopMenuItemLink > DesktopMenuItemColumn 1`] = `
<div>
  <nav
    class="hidden items-center gap-x-12 px-8 lg:flex"
    role="navigation"
  >
    <div>
      <a
        aria-controls=":rb:"
        aria-haspopup="true"
        class="flex button button-outline button-black button-pill button-medium !border-0 cursor-pointer"
        data-name="ButtonAnchor"
        href="/l/discover-club-med"
        id=":ra:"
      >
        <span
          class="flex items-center justify-center gap-x-8 w-full"
        >
          Discover
        </span>
      </a>
      <div
        class="absolute inset-x-0 top-full bg-white px-20 py-40 hidden"
      >
        <div
          aria-labelledby=":ra:"
          class="mb-24 flex columns-5 justify-center gap-x-40"
          id=":rb:"
          role="menu"
        >
          <div
            class="break-inside-avoid-column space-y-32 text-b3"
          >
            <div
              role="group"
            >
              <a
                class="block font-bold"
                href="/s"
                role="menuitem"
              >
                Our products
              </a>
              <ul
                class="mt-20 space-y-8"
                role="group"
              >
                <li
                  role="presentation"
                >
                  <a
                    href="/o/all-inclusive-sun-holidays"
                    role="menuitem"
                  >
                    All inclusive sun holidays
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;

exports[`<DesktopMenu /> > renders DesktopMenu > DesktopMenu 1`] = `
<div>
  <nav
    class="hidden items-center gap-x-12 px-8 lg:flex"
    role="navigation"
  >
    <div>
      <a
        aria-controls=":r1:"
        aria-haspopup="true"
        class="flex button button-outline button-black button-pill button-medium !border-0 cursor-pointer"
        data-name="ButtonAnchor"
        href="/l/discover-club-med"
        id=":r0:"
      >
        <span
          class="flex items-center justify-center gap-x-8 w-full"
        >
          Discover
        </span>
      </a>
      <div
        class="absolute inset-x-0 top-full bg-white px-20 py-40 hidden"
      >
        <div
          aria-labelledby=":r0:"
          class="mb-24 flex columns-5 justify-center gap-x-40"
          id=":r1:"
          role="menu"
        >
          item
        </div>
      </div>
    </div>
  </nav>
</div>
`;

exports[`<DesktopMenu /> > renders DesktopMenuItemColumn > DesktopMenuItemColumn 1`] = `
<div>
  <nav
    class="hidden items-center gap-x-12 px-8 lg:flex"
    role="navigation"
  >
    <div>
      <a
        aria-controls=":r9:"
        aria-haspopup="true"
        class="flex button button-outline button-black button-pill button-medium !border-0 cursor-pointer"
        data-name="ButtonAnchor"
        href="/l/discover-club-med"
        id=":r8:"
      >
        <span
          class="flex items-center justify-center gap-x-8 w-full"
        >
          Discover
        </span>
      </a>
      <div
        class="absolute inset-x-0 top-full bg-white px-20 py-40 hidden"
      >
        <div
          aria-labelledby=":r8:"
          class="mb-24 flex columns-5 justify-center gap-x-40"
          id=":r9:"
          role="menu"
        >
          <div
            class="break-inside-avoid-column space-y-32 text-b3"
          >
            Column
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
